vapoursynth: add withPlugins interface

sbruder opened the pull request:
This PR patches vapoursynth to allow specifying the plugin directory in an environment variable. This allows for a `withPlugins` interface to create a vapoursynth environment with the specified plugins.

###### Motivation for this change

Currently, there is no nice way for using vapoursynth with plugins. A workaround exists (https://github.com/NixOS/nixpkgs/pull/58859#issuecomment-479397341), but has its drawbacks (e.g. temporary usage with `nix run` or `nix-shell` does not work, it does not support multiple environments with different plugins etc.).

Since there is no upstream support for specifying a plugin directory other than in the global configuration file, I added the ability to load plugins from the directory specified in `VAPOURSYNTH_PLUGIN_PATH` as a patch. This is probably not the best option for implementing this, but my other idea (resolving the plugin path relative to the shared library at runtime) would require copying (instead of symlinking) the vapoursynth shared library so the plugin path can be found relative to it. That seemed even worse, so I chose the former option.

An example plugin environment can be built from the following expression: `vapoursynth.withPlugins [ vapoursynth-mvtools vapoursynth-editor ]`  
This includes `vapoursynth-editor` as a plugin (even though it isn’t), since that way the plugin path also gets added to the wrapper of `vsedit`.

This also adds the `VAPOURSYNTH_PLUGIN_PATH` environment variable to the mpv wrapper when vapoursynth support is enabled. Example:

```nix
wrapMpv
  (mpv-unwrapped.override {
    vapoursynthSupport = true;
    vapoursynth = vapoursynth.withPlugins [
      vapoursynth-mvtools
    ];
  })
  { }
```

I’m open for suggestions, as I acknowledge that this solution is somewhat of a hack.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sbruder committed:
maintainers: add sbruder

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110365` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>vapoursynth-editor: log <a href="https://termbin.com/tmmb">https://termbin.com/tmmb</a></li>
  </ul>
</details>
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>vapoursynth</li>
<li>vapoursynth-mvtools</li>
</ul>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110365` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>vapoursynth</li>
<li>vapoursynth-editor</li>
<li>vapoursynth-mvtools</li>
</ul>
</details>

sbruder commented:
Force pushed to rebase and to fix dependencies (`propagatedBuildInputs`) of plugins not being included in the python path.

rnhmjoj reviewed (commented):
Thank you for this! I have just tested it with mpv and it's working great.
I'm just not entirely convinced by the `buildEnv` workaround: I'll try asking around if there's a better way.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110365` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>vapoursynth-editor: log <a href="https://termbin.com/t454">https://termbin.com/t454</a></li>
  </ul>
</details>
<details>
<summary>
1 package built:
</summary>
<ul>
<li>vapoursynth</li>
</ul>
</details>

rnhmjoj commented:
Ok, I managed to simplify it a bit. I think `symlinkJoin` instead of `buildEnv` is more appropriate, because what we really want here are real directories of symlinked files. Here's an updated patch

```patch
commit cd7026359549cf0ee6efa7de256c52e31a80e3de
Author: Simon Bruder <simon@sbruder.de>
Date:   Thu Jan 21 15:53:58 2021 +0100

    vapoursynth: add withPlugins interface
    
    Since nixpkgs does not currently have an attribute set with plugins, the
    usage differes from python-style withPackages as it accepts a list
    instead of a function as a parameter.
    
    Using plugins in other programs that require vapoursynth (like `vsedit`)
    can be achieved by adding it to the plugin list (e.g.
    `vapoursynth.withPlugins [ vapoursynth-editor … ]`) or by passing the
    `VAPOURSYNTH_PLUGIN_PATH` variable.

diff --git a/pkgs/development/libraries/vapoursynth/default.nix b/pkgs/development/libraries/vapoursynth/default.nix
index 93f8d3c5ae5..6723868746d 100644
--- a/pkgs/development/libraries/vapoursynth/default.nix
+++ b/pkgs/development/libraries/vapoursynth/default.nix
@@ -1,4 +1,6 @@
-{ lib, stdenv, fetchFromGitHub, pkg-config, autoreconfHook, makeWrapper
+{ lib, stdenv, fetchFromGitHub, pkg-config, autoreconfHook
+, symlinkJoin, makeWrapper
+, vapoursynth
 , zimg, libass, python3, libiconv
 , ApplicationServices
 , ocrSupport ?  false, tesseract ? null
@@ -21,6 +23,10 @@ stdenv.mkDerivation rec {
     sha256 = "1krfdzc2x2vxv4nq9kiv1c09hgj525qn120ah91fw2ikq8ldvmx4";
   };
 
+  patches = [
+    ./plugin-path-environment-variable.patch
+  ];
+
   nativeBuildInputs = [ pkg-config autoreconfHook makeWrapper ];
   buildInputs = [
     zimg libass
@@ -42,6 +48,30 @@ stdenv.mkDerivation rec {
     # wrapper, what python3 version was used to build vapoursynth so
     # the right python3.sitePackages will be used there.
     inherit python3;
+
+    withPlugins = plugins:
+      let
+        pythonEnvironment = python3.buildEnv.override { extraLibs = plugins; };
+      in
+      symlinkJoin {
+        name = "${vapoursynth.name}-with-plugins";
+        paths = [ vapoursynth ] ++ plugins;
+        buildInputs = [ makeWrapper ];
+        postBuild = ''
+          for symlink in $out/bin/*; do
+            # remove the original symlink
+            original="$(realpath "$symlink")"
+            rm "$symlink"
+
+            makeWrapper "$original" "$symlink" \
+              --set VAPOURSYNTH_PLUGIN_PATH $out/lib/vapoursynth \
+              --prefix PYTHONPATH : ${pythonEnvironment}/${python3.sitePackages}
+          done
+        '';
+
+        # passthrus must be specified here, again
+        passthru = { inherit python3; };
+      };
   };
 
   postInstall = ''
@@ -54,7 +84,6 @@ stdenv.mkDerivation rec {
     homepage    = "http://www.vapoursynth.com/";
     license     = licenses.lgpl21;
     platforms   = platforms.x86_64;
-    maintainers = with maintainers; [ rnhmjoj tadeokondrak ];
+    maintainers = with maintainers; [ rnhmjoj sbruder tadeokondrak ];
   };
-
 }
diff --git a/pkgs/development/libraries/vapoursynth/plugin-path-environment-variable.patch b/pkgs/development/libraries/vapoursynth/plugin-path-environment-variable.patch
new file mode 100644
index 00000000000..80f98633d7f
--- /dev/null
+++ b/pkgs/development/libraries/vapoursynth/plugin-path-environment-variable.patch
@@ -0,0 +1,16 @@
+diff --git a/src/core/vscore.cpp b/src/core/vscore.cpp
+index 2d29844d..fd4cbea5 100644
+--- a/src/core/vscore.cpp
++++ b/src/core/vscore.cpp
+@@ -1351,6 +1351,11 @@ VSCore::VSCore(int threads) :
+     } // If neither exists, an empty string will do.
+ #endif
+ 
++    const char *pluginPath = getenv("VAPOURSYNTH_PLUGIN_PATH");
++    if (pluginPath) {
++        loadAllPluginsInPath((std::string)pluginPath, filter);
++    }
++
+     VSMap *settings = readSettings(configFile);
+     const char *error = vs_internal_vsapi.getError(settings);
+     if (error) {

```

rnhmjoj commented:
I tested `symlinkJoin`, as before, with mpv and mvtools and everything looks good.

tadeokondrak committed:
vapoursynth: add withPlugins interface

Co-authored-by: Simon Bruder <simon@sbruder.de>

tadeokondrak committed:
vapoursynth-editor: allow adding plugins without rebuilding

tadeokondrak commented:
Here's my attempt at this which I think is more reliable, because it modifies VapourSynth's shared library:

https://github.com/tadeokondrak/nixpkgs/commit/1f00195d337fc330469f8b54c5f3c64b5480b689

I also changed VSEdit to allow adding plugins without rebuilding:

https://github.com/tadeokondrak/nixpkgs/commit/d35511d82ab07cf1d5f22a62b76061c9cfc2e6f1

sbruder commented:
Thanks @rnhmjoj for the suggestions. That looks cleaner than my workaround.

@tadeokondrak: Your attempt looks very nice, since it doesn’t rely on hacks (environment variables) like mine, thank you! It worked when I tried it with a basic setup in vsedit, however I didn’t yet have the time to test it with mpv or in more depth. Do you want to open a new PR for your implementation?

rnhmjoj commented:
> Here's my attempt at this which I think is more reliable, because it modifies VapourSynth's shared library:

Indeed, but it's more complex: I don't think I can review this.

Anyway, since plugins path can be configured via `SystemPluginDir`, would it be possible to patch vapoursynth to load a custom generated vapoursynth.conf that points to $out? This would not require an environment variable or adding new code.

sbruder commented:
I rebased onto master and incorporated @rnhmjoj’s suggestions.

I looked into @tadeokondrak’s attempt again. `vspipe` in the vapoursynth package isn’t wrapped, but when manually passing `LD_LIBRARY_PATH`, it works. With that, mpv and plain python scripts that import vapoursynth also work.

I agree that using a preload is complex, but given vapoursynth’s limited ability to load plugins from dynamic locations, it seems like any solution requires at least some new code.  
It is relatively easy to patch vapoursynth to use a static configuration file in `$out`, but if I understand it correctly, that would require recompiling vapoursynth every time the plugins change. Since vapoursynth uses a hardcoded path to the configuration file (https://github.com/vapoursynth/vapoursynth/blob/d2f6cb0294e725605e8d86596172f3e638de7435/src/core/vscore.cpp#L1339-L1352), using that mechanism to load plugins from a configuration file outside of vapoursynth’s store path would also require adding quite a bit of new code (at least from my understanding, please correct me if I’m mistaken).

rnhmjoj commented:
> I rebased onto master and incorporated @rnhmjoj’s suggestions.

Thanks

> it seems like any solution requires at least some new code. 

Yes, I came to the same conclusion, unfortunately.

> using that mechanism to load plugins from a configuration file outside
> of vapoursynth’s store path would also require adding quite a bit of
> new code (at least from my understanding, please correct me if I’m
> mistaken).

No, you're right: mine wasn't a good idea afterwall :)

Both yours and @tadeokondrak's solutions look good to me, I don't have a
strong preference for either one.

On 05-02-21, Simon Bruder wrote:
> I rebased onto master and incorporated @rnhmjoj’s suggestions.
> 
> I looked into @tadeokondrak’s attempt again. `vspipe` in the vapoursynth package isn’t wrapped, but when manually passing `LD_LIBRARY_PATH`, it works. With that, mpv and plain python scripts that import vapoursynth also work.
> 
> I agree that using a preload is complex, but given vapoursynth’s limited ability to load plugins from dynamic locations, it seems like any solution requires at least some new code.  
> It is relatively easy to patch vapoursynth to use a static configuration file in `$out`, but if I understand it correctly, that would require recompiling vapoursynth every time the plugins change. Since vapoursynth uses a hardcoded path to the configuration file (https://github.com/vapoursynth/vapoursynth/blob/d2f6cb0294e725605e8d86596172f3e638de7435/src/core/vscore.cpp#L1339-L1352), using that mechanism to load plugins from a configuration file outside of vapoursynth’s store path would also require adding quite a bit of new code (at least from my understanding, please correct me if I’m mistaken).
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixpkgs/pull/110365#issuecomment-774302594

sbruder commented:
Sorry for the wrong pings/review requests, something went wrong while rebasing.

sbruder commented:
I swiched to @tadeokondrak’s approach, wrapped `vspipe` and added the needed makeWrapper options to mpv.

Also, to make `propagatedBuildInputs` work, I added a function that recursively gets the `propagatedBuildInputs` of all plugins and added the python environment as path to the environment. I’m not sure if this is the best way to do it.

tadeokondrak commented on pkgs/development/libraries/vapoursynth/default.nix:
```diff
@@ -42,6 +47,60 @@ stdenv.mkDerivation rec {
+
+      ext = stdenv.targetPlatform.extensions.sharedLibrary;
+    in
+    buildEnv {
+      name = "${vapoursynth.name}-with-plugins";
+      paths = [ vapoursynth pluginLoader pythonEnvironment ];
+      buildInputs = [ makeWrapper patchelf ];
+      postBuild = ''
+        rm $out/lib/libvapoursynth${ext}
+        cp ${vapoursynth}/lib/libvapoursynth${ext} $out/lib/libvapoursynth${ext}
+        chmod +w $out/lib/libvapoursynth${ext}
+        patchelf $out/lib/libvapoursynth${ext} \
+            --add-needed libvapoursynth-nix-plugins${ext}
+        chmod -w $out/lib/libvapoursynth${ext}
+
+        rm -r $out/bin
+        makeWrapper ${vapoursynth}/bin/vspipe $out/bin/vspipe \
+            --prefix LD_LIBRARY_PATH : $out/lib \
+            --prefix PYTHONPATH : $out/${python3.sitePackages}
+      '';
```
```suggestion
    runCommand "${vapoursynth.name}-with-plugins" {
      nativeBuildInputs = [ makeWrapper ];
    } ''
      mkdir -p \
        $out/bin \
        $out/lib/pkgconfig \
        $out/lib/vapoursynth \
        $out/${python3.sitePackages}

      for textFile in \
          lib/pkgconfig/vapoursynth{,-script}.pc \
          lib/libvapoursynth.la \
          lib/libvapoursynth-script.la \
          ${python3.sitePackages}/vapoursynth.la
      do
          substitute ${vapoursynth}/$textFile $out/$textFile \
              --replace "${vapoursynth}" "$out"
      done

      for binaryFile in \
          lib/libvapoursynth${ext} \
          lib/libvapoursynth-script${ext}.0.0.0
      do
        old_rpath=$(patchelf --print-rpath ${vapoursynth}/$binaryFile)
        new_rpath="$old_rpath:$out/lib"
        patchelf \
            --set-rpath "$new_rpath" \
            --output $out/$binaryFile \
            ${vapoursynth}/$binaryFile
        patchelf \
            --add-needed libvapoursynth-nix-plugins${ext} \
            $out/$binaryFile
      done

      for binaryFile in \
          ${python3.sitePackages}/vapoursynth${ext} \
          bin/.vspipe-wrapped
      do
          old_rpath=$(patchelf --print-rpath ${vapoursynth}/$binaryFile)
          new_rpath="''${old_rpath//"${vapoursynth}"/"$out"}"
          patchelf \
              --set-rpath "$new_rpath" \
              --output $out/$binaryFile \
              ${vapoursynth}/$binaryFile
      done

      ln -s \
          ${pluginLoader}/lib/libvapoursynth-nix-plugins${ext} \
          $out/lib/libvapoursynth-nix-plugins${ext}
      ln -s ${vapoursynth}/include $out/include
      ln -s ${vapoursynth}/lib/vapoursynth/* $out/lib/vapoursynth
      ln -s \
          libvapoursynth-script${ext}.0.0.0 \
          $out/lib/libvapoursynth-script${ext}
      ln -s \
          libvapoursynth-script${ext}.0.0.0 \
          $out/lib/libvapoursynth-script${ext}.0

      makeWrapper $out/bin/.vspipe-wrapped $out/bin/vspipe \
          --prefix PYTHONPATH : $out/${python3.sitePackages}
    '';
```

I think having to wrap vspipe shows that this solution was actually incomplete. Any other binary depending on libvapoursynth-script would have to be wrapped in this way, but this change shouldn't require changes in other packages.

I think this new build gets replaces all the vapoursynth references now.

sbruder replied:
Thanks! I applied the changes and re-added support for python plugins.

tadeokondrak commented:
I asked for feedback from someone else and they asked why the plugin loader doesn't directly use the path to the wrapper store directory. This is because that would be a circular dependency, but it made me think that we should at least link all the plugins to `$out/lib/vapoursynth`.

```diff
diff --git a/pkgs/development/libraries/vapoursynth/default.nix b/pkgs/development/libraries/vapoursynth/default.nix
index 2bc0d00b521..70fd297610a 100644
--- a/pkgs/development/libraries/vapoursynth/default.nix
+++ b/pkgs/development/libraries/vapoursynth/default.nix
@@ -1,5 +1,5 @@
 { lib, stdenv, fetchFromGitHub, pkg-config, autoreconfHook, makeWrapper
-, runCommandCC, runCommand, vapoursynth, writeText, patchelf
+, runCommandCC, runCommand, vapoursynth, writeText, patchelf, buildEnv
 , zimg, libass, python3, libiconv
 , ApplicationServices
 , ocrSupport ?  false, tesseract ? null
@@ -60,6 +60,12 @@ stdenv.mkDerivation rec {
 
       deepPlugins = plugins ++ (getRecursivePropagatedBuildInputs plugins);
 
+      pluginsEnv = buildEnv {
+        name = "vapoursynth-plugins-env";
+        pathsToLink = [ "/lib/vapoursynth" ];
+        paths = deepPlugins;
+      };
+
       pluginLoader = let
         source = writeText "vapoursynth-nix-plugins.c" ''
           void VSLoadPluginsNix(void (*load)(void *data, const char *path), void *data) {
@@ -102,6 +108,10 @@ stdenv.mkDerivation rec {
               --replace "${vapoursynth}" "$out"
       done
 
+      for binaryPlugin in ${pluginsEnv}/lib/vapoursynth/*; do
+          ln -s $binaryPlugin $out/''${binaryPlugin#"${pluginsEnv}/"}
+      done
+
       for pythonPlugin in ${pythonEnvironment}/${python3.sitePackages}/*; do
           ln -s $pythonPlugin $out/''${pythonPlugin#"${pythonEnvironment}/"}
       done
```

sbruder commented:
That seems like a good idea. I applied your patch.

rnhmjoj commented:
I tested again today and everything seems in order. I would perhaps move the code to an external expression and do `withPlugins = import ./plugins-interface.nix` to keep the main package shorter.

sbruder commented:
Splitting the function into its own file is a good idea considering it amounts for more than half of the code. I rebased and incorporated this suggestion.

rnhmjoj commented on pkgs/development/libraries/vapoursynth/default.nix:
```diff
@@ -36,12 +41,17 @@ stdenv.mkDerivation rec {
     inherit python3;
+
+    withPlugins = import ./plugin-interface.nix {
+      inherit lib python3 buildEnv writeText concatMapStringsSep runCommandCC
```
`concatMapStringsSep` is inside `lib`, you can avoid passing it separately if you use `lib.concatMapStringsSep`.

sbruder replied:
Thanks. I fixed it.

rnhmjoj commented:
Looks good to me. @tadeokondrak?

tadeokondrak reviewed (approved):
Tested with vspipe and vsedit again to be sure, looks good and works well.

sbruder commented:
Thanks for helping me with this PR and merging it!
